### PR TITLE
Upgrade Web Server to use Indy VDR

### PR DESCRIPTION
**Pull request by AlexanderShenshin**
This PR contains `Indy SDK` -> `Indy VDR` migration for VON network Web Server and includes changes adopted from previous work in this direction (see [fork](https://github.com/andrewwhitehead/von-network/tree/indy-vdr)).

### CODE REVIEW

**Comment by AlexanderShenshin** [comment]
@swcurran, @WadeBarnes ^

File: scripts/start_webserver.sh [hunk 1]
```diff
@@ -2,20 +2,13 @@
...
-		von_generate_transactions -s "$IPS"
-	elif [ ! -z "$IP" ]; then
-		echo von_generate_transactions -i "$IP"
-		von_generate_transactions -i "$IP"
-	else
-		echo von_generate_transactions
-		von_generate_transactions
-	fi
+if [ ! -f "/home/indy/config/genesis.txn" ] && [ -z "${GENESIS_URL}" ] && [ -z "${GENESIS_FILE}" ]; then
+  bash ./scripts/init_genesis.sh
```
Comment (WadeBarnes) [inline scripts/start_webserver.sh:6]: The `start_webserver.sh` script, or any of the scripts, do not exist on the web server base image due to how the base images were separated.

File: server/anchor.py [hunk 2]
```diff
@@ -175,52 +171,46 @@ class NotReadyException(AnchorException):
...
-            self._pool = await pool.open_pool_ledger(pool_name, json.dumps(pool_cfg))
-        except IndyError as e:
-            raise AnchorException("Error opening pool ledger connection") from e
+                LOGGER.info("Finished pool refresh: %s", self._pool.last_status)
+            except VdrError as e:
+                if e.code == VdrErrorCode.POOL_TIMEOUT and attempts < 5:
+                    LOGGER.info("Pool timeout occurred, waiting to retry")
+                    attempts += 1
+                    await asyncio.sleep(10)
+                    continue
```
Comment (WadeBarnes) [inline server/anchor.py:210]: I like this!  Please make the number of attempts and delay between attempts configurable.

File: docker/Dockerfile.server [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+ENV RUST_LOG ${RUST_LOG:-warning}
+
+ADD server/requirements.txt server/
+RUN pip3 install -U pip && \
+    pip install --no-cache-dir -r server/requirements.txt
+
+ADD --chown=indy:indy config ./config
+ADD --chown=indy:indy server ./server
+
+CMD python3 -m server.server
```
Comment (WadeBarnes) [inline docker/Dockerfile.server:13]: The change to the startup `cmd` complicates the genesis file management a bit.

**Review by WadeBarnes — CHANGES_REQUESTED** [review]
Over all the changes look good.  I have a few general comments here with some specifics below.

### Separation of the Node and Web Server Images
In general I can see why you'd want to separate them, but it does cause some complications.  Additionally it would require updates to the GHA Workflows that build and publish the base image(s).  Having a single base image allows for more consistency across the various scripts and also in how the genesis files are managed and kept in sync, even though it introduces unused artifacts on the image depending on its use case.

### Genesis File Management
How the genesis files are managed has changed with the separation of the images and the addition of the use of the `config/genesis.txn` file.  Adding this additional file complicates genesis file management a bit and actually opens the door for the genesis files used by the nodes and the webserver to inadvertently to get out of sync.

### Broken manage command - generateDid
The generateDid command was inadvertently broken by the separation of the base images.

**Comment by AlexanderShenshin** [comment]
Makes sense, these are good points. Thank you for detailed feedback! 
Will be fixed soon.

**Comment by AlexanderShenshin** [comment]
@WadeBarnes I pushed PR fixes:
- Reverted separation of Node and Web Server Docker images
- Reverted changes to genesis file management (different paths/scripts), but kept separate `init_genesis` script and updated all startup scripts to use it (to reduce code duplication).
- Added `POOL_CONNECTION_ATTEMPTS` and `POOL_CONNECTION_DELAY` ENV variables for pool connection retries.

File: server/anchor.py (hunk 2 above)
Comment (AlexanderShenshin) [reply server/anchor.py]: Done.

File: scripts/start_webserver.sh (hunk 1 above)
Comment (AlexanderShenshin) [reply scripts/start_webserver.sh]: Fixed by reverting base images separation.

File: scripts/start_synctest.sh [hunk 4]
```diff
@@ -2,18 +2,11 @@
...
-	else
-		echo von_generate_transactions
-		von_generate_transactions
-	fi
+if [ ! -f "/home/indy/config/genesis.txn" ] && [ -z "${GENESIS_FILE}" ]; then
+  bash ./scripts/init_genesis.sh
 fi
 
+pip3 install -U pip &&
+pip install --no-cache-dir -r server/requirements.txt
```
[the remark this answers is not in the document]
Comment (AlexanderShenshin) [reply scripts/start_synctest.sh]: Fixed by reverting base images separation.

File: manage [hunk 5]
```diff
@@ -886,10 +898,12 @@ case "${COMMAND}" in
       remove_taa
     ;;
   build)
-      docker build $(initDockerBuildArgs) -t von-network-base .
+      docker-compose build webserver
+      docker build $(initDockerBuildArgs) -t von-network-base . -f docker/Dockerfile.node
     ;;
   rebuild)
-      docker build --no-cache --progress plain $(initDockerBuildArgs) -t von-network-base .
+      docker-compose build webserver
```
[the remark this answers is not in the document]
Comment (AlexanderShenshin) [reply manage]: Not needed as we reverted base images separation.

File: manage [hunk 6]
```diff
@@ -886,10 +898,12 @@ case "${COMMAND}" in
       remove_taa
     ;;
   build)
-      docker build $(initDockerBuildArgs) -t von-network-base .
+      docker-compose build webserver
```
[the remark this answers is not in the document]
Comment (AlexanderShenshin) [reply manage]: Not needed as we reverted base images separation.

File: .gitignore [hunk 7]
```diff
@@ -148,13 +148,17 @@ Temporary Items
 
 Pipfile
 
-# Custom TAA
+# Config files (custom TAA, genesis)
 config/aml.json
 config/taa.json
 config/.samples.used
+config/genesis.txn
```
[the remark this answers is not in the document]
Comment (AlexanderShenshin) [reply .gitignore]: Removed as we reverted genesis file management changes.

File: docker-compose.yml [hunk 8]
```diff
@@ -45,10 +45,8 @@ services:
     ports:
       - ${WEB_SERVER_HOST_PORT:-9000}:8000
     volumes:
-      - ./config:/home/indy/config
-      - ./server:/home/indy/server
-      - webserver-cli:/home/indy/.indy-cli
-      - webserver-ledger:/home/indy/ledger
```
[the remark this answers is not in the document]
Comment (AlexanderShenshin) [reply docker-compose.yml]: Done.

File: scripts/init_genesis.sh [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    von_generate_transactions -s "$IPS"
+elif [ ! -z "$IP" ]; then
+    echo von_generate_transactions -i "$IP"
+    von_generate_transactions -i "$IP"
+else
+    echo von_generate_transactions
+    von_generate_transactions
+fi
+
+cp /home/indy/ledger/sandbox/pool_transactions_genesis /home/indy/config/genesis.txn
```
[the remark this answers is not in the document]
Comment (AlexanderShenshin) [reply scripts/init_genesis.sh]: This separate script now used just to reduce code duplication in startup script and uses the same genesis file management approach as before.

File: docker/Dockerfile.server (hunk 3 above)
Comment (AlexanderShenshin) [reply docker/Dockerfile.server]: Removed as we reverted base images separation.

File: server/anchor.py [hunk 10]
```diff
@@ -63,35 +63,52 @@
...
+        else:
+            if "result" in reply:
+                data = reply["result"]["data"]
+                data["Node_info"]["Name"] = node
+            elif "reason" in reply:
+                data = {"name": node, "error": reply["reason"]}
+            else:
+                data = {"name": node, "error": "unknown error"}
+        ret.append(data)
+    return ret
```
Comment (WadeBarnes) [inline server/anchor.py:103]:
```suggestion
def format_validator_info(node_data):
    node_aliases = list(node_data.keys())
    node_aliases.sort()

    ret = []
    for node in node_aliases:
        try:
            reply = json.loads(node_data[node])
        except json.JSONDecodeError:
            data = {"Node_info": {"Name": node}, "error": node_data[node]}  # likely 'timeout'
        else:
            if "result" in reply:
                data = reply["result"]["data"]
            elif "reason" in reply:
                data = {"Node_info": {"Name": node}, "error": reply["reason"]}
            else:
                data = {"Node_info": {"Name": node}, "error": "unknown error"}
        ret.append(data)
    
    return ret
```

The name of the node needs to be placed into `Node_info.Name` for `/status/text` to work when there is an error result.

Updates are also required to `app.methods.formatValidatorStatus` in `server/static/include/js/index.js` to handle the error results.

Adding:
```
          if (node.error)
            continue
```
to the `for` loop will retain the existing display functionality when nodes encounter an error condition.  In the following example Node 2 was stopped:

image

**Review by WadeBarnes — CHANGES_REQUESTED** [review]
Looking good.  That cleaned things up a fair amount.

A couple small issues discovered and detailed below.

File: server/anchor.py (hunk 10 above)
Comment (AlexanderShenshin) [reply server/anchor.py]: Good catch. Done.

**Review by WadeBarnes — COMMENTED** [review]
@AlexanderShenshin, Looks good to me.  Do you want to clean-up any of the commit history before I approve and merge, or do you want to leave all of the commits as they are?

**Comment by AlexanderShenshin** [comment]
@WadeBarnes I pushed some cleanups for commit names, I think that's it.
Btw, I don't mind if we use squash if needed.

**Review by WadeBarnes — APPROVED** [review]

**Pull request merged**